### ISSUE TITLE
How to debug TLS/SSL Web errors?

### ISSUE THREAD

**Issue by jackyaz** [netdata/netdata#6453]
##### Question summary
How to debug TLS/SSL Web errors? I've put cert and key in place, ensured netdata user/group has read access, but browsers cannot connect using https

##### OS / Environment
Lubuntu 18.04

**Comment by Ferroin** [comment]
Low-hanging fruit first:
* Double check that you're specifying the correct port for the connection in the URL.  Not likely this is an issue, but if you had Netdata set up to serve it's web interface over port 80 and weren't specifying a port when connecting, then HTTPS wouldn't work because the browser would be trying port 443 instead.
* If you're using a self-signed certificate, you will probably have to explicitly import it as a trusted certificate in the browsers you're using for the connection to work without complaining about a TLS/SSL error.

As far as further debugging:

The error logs for Netdata should show if it's unable to load the certificate and/or key, but given what you've said, that's pretty unlikely to be the cause.  I don't know if it will actually show any useful info beyond that if there are connection issues.

For further debugging, `openssl s_client` is a wonderful tool for checking if TLS is working correctly (and it should be available on pretty much any Linux distribution out of the box).  General syntax is `openssl s_client -connect <hostname>:<port>`.  What that will do is attempt to establish a TLS connection to `<hostname>` on `<port>` and log all the little details about the certificate, the TLS handshake, and any of the other related info on the connection,, then let you talk to the server directly through the terminal (if you're just trying to check that TLS is set up correctly, you can just hit Ctrl-D or Ctrl-C at this point to exit).

**Comment by jackyaz** [comment]
Configured to listen on port 9999 - HTTP works but HTTPS coughs up protocol error (which leads me to assume it isn't listening for an SSL-enabled connection

Certificate CA is trusted on my machines (I use lots of self-signed certs from the same CA without issue)

I'll fire up openssl and see what I get back

**Comment by thiagoftsm** [comment]
Hi @jackyaz,

When there is a problem to load the library or to load the certificate, you will have one of the following error message in your error.log:

- "SSL library cannot be initialized."
- "Cannot create a new SSL context, netdata won't encrypt communication"
- "To use encryption it is necessary to set "ssl certificate" and "ssl key" in [web] !\n"

Do you see any of these errors in your error.log?
Another thing that comes in my mind, it is about a possible problem with the OpenSSL version, during our development time, I saw that OpenSSL changed the functions necessary to set it up, no less important some options that were available for oldest version are not more present in the newest. What is your OpenSSL version?

**Comment by jackyaz** [comment]
OpenSSL 1.1.1  11 Sep 2018

**Comment by jackyaz** [comment]
No SSL errors when watching log:

tail -F /var/log/netdata/error.log | grep -i *SSL*

**Comment by thiagoftsm** [comment]
What is the protocol error that you are having?
I remember when I was developing with the OpenSSL I got some errors due one of the following motives:

1 - TLS 1.3 was not accepted by all the browsers, so we decided to move in front with TLS 1.2
2 - SSLv3 was not a problem in Centos 7, that has OpenSSL 0.9.2, but it was a problem with the newest Linux/BSD
3 - I created a self signed certificate that needed a password to start, due the wrong parameters given to OpenSSL.

Best Regards!

**Comment by jackyaz** [comment]
This site can’t provide a secure connection 10.14.16.99 sent an invalid response.
Try running Windows Network Diagnostics.
ERR_SSL_PROTOCOL_ERROR

**Comment by jackyaz** [comment]
[web]
        default port = 9999
        bind to = 10.14.16.99=dashboard|registry|badges|management|streaming|netdata.conf
        ssl key = /etc/netdata/ssl/netdata.apps.pem
        ssl certificate = /etc/netdata/ssl/netdata.apps.crt

**Comment by Ferroin** [comment]
@jackyaz You wouldn't happen to be using Internet Explorer, would you?  Because if so, I suspect that to be the issue.  I don't think it ever got support for TLS 1.1, let alone TLS 1.2.

**Comment by jackyaz** [comment]
Nope, Google Chrome. OpenSSL reports:

```
jack@DockerBox:~$ openssl s_client -connect 10.14.16.99:9999
CONNECTED(00000005)
140479642710464:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../ssl/record/ssl3_record.c:332:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 5 bytes and written 313 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---

```
Entirely possible I've configured something wrong, but I'm missing where!

**Comment by thiagoftsm** [comment]
I will write only one comment here to answer all the last comments that you wrote.

The error "ERR_SSL_PROTOCOL_ERROR" from Google Chrome means a lot of different things, but fortunately you also gave us the openssl client output, that means only one thing. Let me do some questions for you, that you help you to confirm the motive openssl cannot get the certificate from the netdata.

1 - Did you compile netdata with support to openssl? You can confirm this running the command :
ldd netdata
2 - Are you sure that the user that is configured your netdata can read and access the directory /etc/netdata/ssl ?
3 - Are the files that you described in "ssl key" and "ssl certificate" owned by the user that runs netdata?

Another thing that I saw is the fact that your ceritificates have different extensions, when we created our ceritifcate to test, both were pem files and we created with the following command:

openssl req -newkey rsa:2048 -nodes -sha512 -x509 -days 365 -nodes -keyout key.pem -out cert.pem

Finally the correct output for openssl command that you ran will be a print of your certificate.

**Comment by jackyaz** [comment]
Hm, you might be on to something:

jack@DockerBox:~$ ldd /usr/sbin/netdata 
	linux-vdso.so.1 (0x00007fffb0dfc000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007faae5f24000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007faae5b86000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007faae5969000)
	libuuid.so.1 => /lib/x86_64-linux-gnu/libuuid.so.1 (0x00007faae5762000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007faae5543000)
	/lib64/ld-linux-x86-64.so.2 (0x00007faae6613000)


The instructions said nothing about compiling it with openssl, however

**Comment by thiagoftsm** [comment]
Hi @jackyaz,

Now it makes more sense, your configuration files were not wrong, so the problem is the fact that your netdata does not support openssl.

**Comment by jackyaz** [comment]
Yes, I suspect the updater hasn't accounted for SSL support being introduced. Though oddly, neither did a git pull and rebuild. I've blown away the install and rebuilding it as we speak

**Comment by jackyaz** [comment]
That solved it. Same thing on my RPi. Is there a way to have the updater check if SSL support has been built?

**Comment by thiagoftsm** [comment]
I am glad with this! 
Are you using a netdata command to update your netdata? Case, yes, I can verify with our team this.

**Comment by jackyaz** [comment]
I used this: https://docs.netdata.cloud/packaging/installer/update/

Which for all intents and purposes sounds like it should have updated with ssl :/

**Comment by thiagoftsm** [comment]
I am bringing our specialist for the talk, because I did not develop the updater, so I cannot answer you properly. @paulkatsoulakis , we solved the problem of @jackyaz today, but he said us that after the update he tried to activate the SSL, but the update was not compiled with SSL. According his suggestion, it would be good for us to have an update that already brings SSL. Can you explained us the whole situation?

**Comment by paulkatsoulakis** [comment]
Hello @jackyaz, could you provide us with the output of /etc/netdata/.environment?
Please be sure to hide any sensitive information, if any is popped out there

**Comment by paulkatsoulakis** [comment]
Just to give a first insight, our kickstart provides dependency management capabilities.
The updater only incorporates running `netdata-installer.sh` with the necessary flags that the initial install was run with.

I guess we could modify our updater to always re-run the dependency management workflow, so that any newly introduced package dependencies are pulled in.  I just wonder though whether we want the updater to take on such a task automatically.

@cakrit @Ferroin thoughts?

**Comment by jackyaz** [comment]
> Hello @jackyaz, could you provide us with the output of /etc/netdata/.environment?
> Please be sure to hide any sensitive information, if any is popped out there

I'll see if I have a backup copy - I've since nuked the installs for a fresh one (via kickstart)

**Comment by paulkatsoulakis** [comment]
I think we got a clear understanding on how things go around.

Adding an optional functionality like open SSL that requires specific packages to be present, it would be a bad practise to trigger our automatic updater to mess with your package management and secretly push new packages in unattended.

I think the best way to handle this is to just make sure that when we announce new functionality we instruct our users to either re-run our kickstart installer or just push the dependencies in and re-run the updater manually.

I would be very interested to know what the community thinks though, i will let this open until we get some more inputs before resolving. In the meantime i will see if we can produce some useful messages through the updater, to help our users get insights on things they should be doing from the updater rather than figuring it out themselves

**Comment by Mesmer88** [comment]
Getting the same error with an install using :  kickstart-base64.sh
I see the binary is statically linked.

Here's the openssl output:

```
CONNECTED(00000003)
140600274642832:error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol:s23_clnt.c:794:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 289 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    Start Time: 1565078861
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---
```

**Comment by paulkatsoulakis** [comment]
Hello @Mesmer88, thank you for reporting this.

Could you confirm the following with us:

1) `uname -a` of your system
2) Output `cat /etc/os-release`
3) Exact command string you used to install netdata using kickstart-static64
4) The complete output of thee kickstart command, because this will help us validate the kind of actions that took place during install

Thanks!

**Comment by Mesmer88** [comment]
Here are the outputs:

1. Uname
```
[root@RAT2VCMSWS03 ~]# uname -a
Linux RAT2VCMSWS03.nvim2.vodafone.com 3.10.0-862.el7.x86_64 #1 SMP Fri Apr 20 16:44:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
2. OS-Release

```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

3. Regarding the install, it's a bit odd. I have to do offline installs, so i download the binary from here:

https://github.com/netdata/netdata/releases/download/v1.16.0/netdata-v1.16.0.tar.gz

then i run the installer with no extra flags.


4. Installer output:

```
Creating directory /opt/netdata
Verifying archive integrity...  100%   All good.
Uncompressing netdata, the real-time performance and health monitoring system  100%  
 --- Attempt to create user/group netdata/netadata --- 
Group 'netdata' already exists.
User 'netdata' already exists.
 --- Add user netdata to required user groups --- 
Group 'docker' does not exist.
 FAILED  Failed to add netdata user to secondary groups 

User 'netdata' is already in group 'nginx'.
Group 'varnish' does not exist.
 FAILED  Failed to add netdata user to secondary groups 

Group 'haproxy' does not exist.
 FAILED  Failed to add netdata user to secondary groups 

User 'netdata' is already in group 'adm'.
Group 'nsd' does not exist.
 FAILED  Failed to add netdata user to secondary groups 

Group 'proxy' does not exist.
 FAILED  Failed to add netdata user to secondary groups 

Group 'squid' does not exist.
 FAILED  Failed to add netdata user to secondary groups 

Group 'ceph' does not exist.
 FAILED  Failed to add netdata user to secondary groups 

User 'netdata' is already in group 'nobody'.
 --- Check SSL certificates paths --- 
 --- Install logrotate configuration for netdata --- 
[/opt/netdata]# chmod 644 /etc/logrotate.d/netdata 
 OK   

 --- Install netdata at system init --- 
Installing systemd service...
[/opt/netdata]# cp system/netdata.service /lib/systemd/system/netdata.service 
 OK   

[/opt/netdata]# systemctl daemon-reload 
 OK   

[/opt/netdata]# systemctl enable netdata 
 OK   

 --- creating quick links --- 
[/opt/netdata]# rm sbin 
 OK   

[/opt/netdata]# ln -s bin sbin 
 OK   

[/opt/netdata/usr]# rm bin 
 OK   

[/opt/netdata/usr]# ln -s ../bin bin 
 OK   

[/opt/netdata/usr]# rm sbin 
 OK   

[/opt/netdata/usr]# ln -s ../bin sbin 
 OK   

[/opt/netdata/usr]# rm local 
 OK   

[/opt/netdata/usr]# ln -s . local 
 OK   

[/opt/netdata]# rm netdata-configs 
 OK   

[/opt/netdata]# ln -s etc/netdata netdata-configs 
 OK   

[/opt/netdata]# rm netdata-web-files 
 OK   

[/opt/netdata]# ln -s usr/share/netdata/web netdata-web-files 
 OK   

[/opt/netdata]# rm netdata-plugins 
 OK   

[/opt/netdata]# ln -s usr/libexec/netdata netdata-plugins 
 OK   

[/opt/netdata]# rm netdata-dbs 
 OK   

[/opt/netdata]# ln -s var/lib/netdata netdata-dbs 
 OK   

[/opt/netdata]# rm netdata-metrics 
 OK   

[/opt/netdata]# ln -s var/cache/netdata netdata-metrics 
 OK   

[/opt/netdata]# rm netdata-logs 
 OK   

[/opt/netdata]# ln -s var/log/netdata netdata-logs 
 OK   

[/opt/netdata/etc/netdata]# rm orig 
 OK   

[/opt/netdata/etc/netdata]# ln -s ../../usr/lib/netdata/conf.d orig 
 OK   

 --- create user config directories --- 
 --- fix permissions --- 
[/opt/netdata]# chmod g+rx\,o+rx /opt 
 OK   

[/opt/netdata]# chown -R netdata:netdata /opt/netdata 
 OK   

 --- fix plugin permissions --- 
[/opt/netdata]# chown root:netdata usr/libexec/netdata/plugins.d/apps.plugin 
 OK   

[/opt/netdata]# chmod 4750 usr/libexec/netdata/plugins.d/apps.plugin 
 OK   

[/opt/netdata]# chown root:netdata usr/libexec/netdata/plugins.d/ioping 
 OK   

[/opt/netdata]# chmod 4750 usr/libexec/netdata/plugins.d/ioping 
 OK   

[/opt/netdata]# chown root:netdata usr/libexec/netdata/plugins.d/cgroup-network 
 OK   

[/opt/netdata]# chmod 4750 usr/libexec/netdata/plugins.d/cgroup-network 
 OK   

[/opt/netdata]# chown root:netdata bin/fping 
 OK   

[/opt/netdata]# chmod 4750 bin/fping 
 OK   

 --- starting netdata --- 
 --- Restarting netdata instance --- 

Stopping all netdata threads
[/opt/netdata]# stop_all_netdata 
Stopping netdata on pid 4762 .....
netdata on pid 4762 stopped.
 OK   

Starting netdata using command 'systemctl start netdata'
[/opt/netdata]# systemctl start netdata 
 OK   


  ^
  |.-.   .-.   .-.   .-.   .-.   .  netdata              .-.   .-.   .-.   .-
  |   '-'   '-'   '-'   '-'   '-'   is installed now!  -'   '-'   '-'   '-'  
  +----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+--->

[/opt/netdata]# chown netdata:netdata /opt/netdata/etc/netdata/netdata.conf 
 OK   

[/opt/netdata]# chmod 0664 /opt/netdata/etc/netdata/netdata.conf 
 OK   
```

**Comment by thiagoftsm** [comment]
Hi @Mesmer88 ,

We will release the solution for this bug with the newest Netdata, until there, you can get the code that is already in the github and use netdata-installer.sh to have this feature enabled.

Best Regards!

**Comment by cakrit** [comment]
> We will release the solution for this bug with the newest Netdata, until there, you can get the code that is already in the github and use netdata-installer.sh to have this feature enabled.

So what's the PR that addressed this? If it's fixed in the master branch, then we should close the issue.

**Comment by thiagoftsm** [comment]
The issue here has direct relationship with #6468, a PR that we already merged, but the kickstart was not recompiled with it yet, this is the motive the users are not receiving the fixes. So I do not know whether we can close it while the kickstart does not carry the fixes.

**Comment by thiagoftsm** [comment]
Hi @Mesmer88 ,

I was talking with @paulkatsoulakis , the person responsible for the kickstart.sh, and we have a doubt, please, can you run the command `netdata -v` to give us the version that you are using of the script?

Best Regards!

**Comment by Mesmer88** [comment]
Here it is:

```
 ./netdata -v
netdata v1.16.0
```

**Comment by cakrit** [comment]
This was fixed in 1.17.0
